src/uniswap.act.md

```diff
@@ -169,12 +180,199 @@ storage
+is:
+```
+87,903,029,871,075,914,254,377,627,908,054,574,944,891,091,886,930,582,284,385,770,809,450,030,037,083
+```
+We don't specify the behaviour of `createPair` after array overflow.
+
+```act
+behaviour createPair-lt of UniswapV2Factory
+interface createPair(address tokenA, address tokenB)
+
+for all
+
+    Pair             : address UniswapV2Pair
+    Length           : uint256
+    Domain_separator : uint256
+    Empty            : address
+    Any              : address
+
+creates storage Pair
+
+    // Set during contract creation
+    #UniswapV2Pair.lockState        |-> 1
+    #UniswapV2Pair.DOMAIN_SEPARATOR |-> Domain_separator
+
+    // Set via initialize
+    #UniswapV2Pair.factory          |-> ACCT_ID
+    #UniswapV2Pair.token0           |-> tokenA
+    #UniswapV2Pair.token1           |-> tokenB
+
+storage
+
+    #UniswapV2Factory.getPair[tokenA][tokenB] |-> Empty  => Pair
+    #UniswapV2Factory.getPair[tokenB][tokenA] |-> 0      => Pair
+    #UniswapV2Factory.allPairs[Length]        |-> Any    => Pair
+    #UniswapV2Factory.allPairs.length         |-> Length => Length + 1
+
+gas
+
+    ( (#if (notBool ( Junk_5 ==K 0 ) ) #then 0 #else 4200 #fi) +Int \
+    ( (#if ( ( Junk_5 ==K 0 ) andBool ( Junk_5 ==K 0 ) ) #then 15000 #else 0 #fi) +Int \
+    ( (#if (notBool ( Junk_6 ==K 0 ) ) #then 0 #else 4200 #fi) +Int \
+    ( (#if ( ( Junk_6 ==K 0 ) andBool ( Junk_6 ==K 0 ) ) #then 15000 #else 0 #fi) +Int \
+    ( (#if ( ( Length ==K ( Length +Int 1 ) ) orBool (notBool ( Junk_8 ==K Length ) ) ) #then 0 #else 4200 #fi) +Int \
+    ( (#if ( ( Junk_8 ==K 0 ) andBool (notBool ( ( Length ==K ( Length +Int 1 ) ) orBool (notBool ( Junk_8 ==K Length ) ) ) ) ) #then 15000 #else 0 #fi) +Int \
+    ( (#if ( ( Any ==K Pair ) orBool (notBool ( Junk_7 ==K Any ) ) ) #then 0 #else 4200 #fi) +Int \
+    ( (#if ( ( Junk_7 ==K 0 ) andBool (notBool ( ( Any ==K Pair ) orBool (notBool ( Junk_7 ==K Any ) ) ) ) ) #then 15000 #else 0 #fi) +Int 2414776 ) ) ) ) ) ) ) )
+
+iff
+
+    tokenA =/= 0
+    tokenA =/= tokenB
+    Empty == 0
+    VCallValue == 0
+
+if
+
+    // The new exchange address
+    Pair == #newAddr(                                 \
+      ACCT_ID,                                        \
+      keccak(                                         \
+        #take(20, #asByteStack(tokenA <<Int 96)) ++   \
+        #take(20, #asByteStack(tokenB <<Int 96))      \
+      ),                                              \
+      UniswapV2Pair_bin                               \
+    )
+
+    // The domain separator is set when the new
+    // exchange is created
+    Domain_separator == keccakIntList(                \
+        Constants.EIP712Domain                        \
+        keccak(#parseByteStackRaw("Uniswap V2"))      \
+        keccak(#parseByteStackRaw("1"))               \
+        VChainId                                      \
+        Pair                                          \
+    )
```

Comment:
would be nice to use a `where` block for these

Reply:
Agree but I was running into rule to long errors using `where` here.